### PR TITLE
Enable shared cookie store

### PR DESCRIPTION
**Pull request by rbuehlma**
Since the removal of cookies.dat (http://github.com/xbmc/xbmc/pull/10611), the CURL implementation of Kodi does not preserve the cookies (over multiple curl sessions) any more which makes it impossible to use in addons where cookies are required (e.g. for identifying the session).

With this PR, I enable the share-interface of curl which allows sharing cookies over multiple curl sessions even in case of multithreading.

Thanks to @ksooo for his time to discuss the solution.

### CODE REVIEW

**Comment by ksooo** [comment]
@FernetMenta @kib do you have an opinion on this?

File: xbmc/filesystem/DllLibCurl.cpp [hunk 1]
```diff
@@ -88,6 +89,25 @@ bool DllLibCurlGlobal::Load()
     return false;
   }
 
+  g_curlShared = share_init();
+  CURLSHcode ret = share_setopt(g_curlShared, CURLSHOPT_LOCKFUNC, DllLibCurlGlobal::lock);
+  if (ret != CURLSHE_OK) {
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: curly bracket has to go into a separate line
same is true for all other places

File: xbmc/filesystem/DllLibCurl.cpp [hunk 2]
```diff
@@ -174,6 +196,29 @@ void DllLibCurlGlobal::CheckIdle()
 #endif
 }
 
+CCriticalSection DllLibCurlGlobal::m_curlLock;
+
+void DllLibCurlGlobal::lock(CURL_HANDLE *handle, curl_lock_data data, curl_lock_access access, void *useptr ) {
+  (void)handle;
+  (void)access;
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: why?

File: xbmc/filesystem/DllLibCurl.cpp [hunk 3]
```diff
@@ -174,6 +196,29 @@ void DllLibCurlGlobal::CheckIdle()
...
+
+  m_curlLock.lock();
+}
+
+void DllLibCurlGlobal::unlock(CURL_HANDLE *handle, curl_lock_data data, void *useptr ) {
+  (void)handle;
+  (void)access;
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: why?

**Review by Rechi — DISMISSED** [review]
coding style had to be fixed, https://codedocs.xyz/xbmc/xbmc/code_guidelines.html#code_guidelines_2

File: xbmc/filesystem/DllLibCurl.cpp (hunk 2 above)
Comment (rbuehlma) [reply xbmc/filesystem/DllLibCurl.cpp]: This was a C&P from a sample. Don't think it is required. -> Removed

**Comment by rbuehlma** [comment]
Fixed code style

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by da-anda** [comment]
I don't know how CURL and out implementation etc works, but might these shared cookies be a security issue? Like, could malicions add-ons read foreign add-on cookies with session IDs and hijack login sessions, or is there something like the "same origin" policiy in place, that browsers have?

**Comment by rbuehlma** [comment]
Are you referring to XSS attacks? That would only work if you can trick CURL into opening a side-resource. As far as I know, CURL does not interpret the returned content, the addon does. So the "same origin" would need the be implemented in the addon from my point of view.

But yes, a malicious addon could probably use the session opened by a different addon but as they run as the same user and are not limited in their functionality, this is already possible now.

**Comment by da-anda** [comment]
Not necessarily XSS, but a malicious add-on using my Amazon login session
to buy stuff via Amazon payments etc.. But if this already is an issue, no
objections to this PR I'm my end. Was just curious. Thanks.

Am 29.08.2017 10:55 vorm. schrieb "rbuehlma" <notifications@github.com>:

Are you referring to XSS attacks? That would only work if you can trick
CURL into opening a side-resource. As far as I know, CURL does not
interpret the returned content, the addon does. So the "same origin" would
need the be implemented in the addon from my point of view.

But yes, a malicious addon could probably use the session opened by a
different addon but as they run as the same user and are not limited in
their functionality, this is already possible now.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/xbmc/xbmc/pull/12728#issuecomment-325599934>, or mute
the thread

**Comment by trummerjo** [comment]
@rbuehlma you should check this pull request: https://github.com/xbmc/xbmc/pull/12737
With this PR you should be able to get the cookie header from the requests, save it, and set the cookie header in further requests (not sure if kodi filters the cookie header in curl!)

**Comment by rbuehlma** [comment]
Thanks for the info. Can we still merge this PR? In my opinion it is a bug if some requests reuse the cookies but not always.

**Comment by wsnipex** [comment]
@arnova @FernetMenta  any comments?

**Comment by FernetMenta** [comment]
shared cookies make no sense and can even interfere in a negative sense. curl is used by many different components and we don't want to share objects between them.
components that require to preserve data, need to do that by themselves

**Pull request closed without merging**